fundamentals/streaming/SynchronousWithNewtonsoftJson/Middleware/SongLyricsMiddleware.cs

```diff
@@ -0,0 +1,59 @@
+using System.IO;
+using System.Threading.Tasks;
+
+using Microsoft.AspNetCore.Http;
+
+using Newtonsoft.Json;
+
+using ASPNetCoreStreamingExample.SynchronousWithNewtonsoftJson.Model;
+
+namespace ASPNetCoreStreamingExample.SynchronousWithNewtonsoftJson.Middleware
+{
+    public class SongLyricsMiddleware
+    {
+        RequestDelegate _next;
+        ILyricsSource _lyricsSource;
+        JsonSerializer _serializer;
+
+        public SongLyricsMiddleware(RequestDelegate next, ILyricsSource lyricsSource, JsonSerializer serializer)
+        {
+            _next = next;
+            _lyricsSource = lyricsSource;
+            _serializer = serializer;
+        }
+
+        public Task InvokeAsync(HttpContext context)
+        {
+            if (!context.Request.Path.StartsWithSegments(new PathString("/middleware/sing")))
+                return _next(context);
+            else
+                return InvokeAsyncImplementation(context);
+        }
+
+        static readonly byte[] JSONArrayStart = new byte[] { (byte)'[' };
+        static readonly byte[] JSONArraySeparator = new byte[] { (byte)',' };
+
+        public async Task InvokeAsyncImplementation(HttpContext context)
+        {
+            await context.Response.Body.WriteAsync(JSONArrayStart, context.RequestAborted);
+
+            while (true)
+            {
+                foreach (string line in _lyricsSource.GetSongLyrics())
+                {
+                    using (var streamWriter = new StreamWriter(context.Response.Body))
+                    using (var jsonWriter = new JsonTextWriter(streamWriter))
```

Comment:
Why re-create these per line?

Reply:
Because we are explicitly writing the array syntax, which isn't going through the JSON text writer's buffering. It could work to ask `JsonTextWriter` to `Flush`, but it feels like it's assuming internal implementation details to use the same instance but interleave its output with external writes.

Reply:
> Why re-create these per line?
> Because we are explicitly writing the array syntax, which isn't going through the JSON text writer's buffering. It could work to ask JsonTextWriter to Flush, but it feels like it's assuming internal implementation details to use the same instance but interleave its output with external writes.
> I have added an explanatory comment.

@Tratcher  review